zebra-consensus/src/block.rs

```diff
@@ -161,17 +161,24 @@ where
             check::coinbase_is_first(&block)?;
             check::subsidy_is_valid(&block, network)?;
 
-            // TODO: context-free header verification: merkle root
+            // Precomputing this avoids duplicating transaction hash computations.
+            let transaction_hashes = block
+                .transactions
+                .iter()
+                .map(|t| t.hash())
+                .collect::<Vec<_>>();
+
+            check::merkle_root_validity(&block, &transaction_hashes)?;
```

Comment:
I'd like to make a change to raise the difficulty of exploiting any validation bugs in Zebra:

Let's check the merkle root before doing any of the transaction checks, to raise the cost of exploiting any vulnerabilities in our transaction checks.

Doing the merkle check right after the difficulty check seems like a good place - that way, we check proof of work, then bind the transactions to it straight away.

Reply:
I'm confused. The change you're describing is what this code already does.

Reply:
In this PR, these header and transaction checks happen between the difficulty check and the merkle root check:
```rust
            let now = Utc::now();
            check::time_is_valid_at(&block.header, now, &height, &hash)
                .map_err(VerifyBlockError::Time)?;
            check::coinbase_is_first(&block)?;
            check::subsidy_is_valid(&block, network)?;
```

Can you move the merkle root check immediately after these difficulty checks?
```rust
            // Do the difficulty checks first, to raise the threshold for
            // attacks that use any other fields.
            check::difficulty_is_valid(&block.header, network, &height, &hash)?;
            check::equihash_solution_is_valid(&block.header)?;
```

Reply:
I see, I thought that you meant the transaction checks, not the block checks. We can move it, but I don't think that it is important to do so.  Building the transaction Merkle tree is more expensive than those checks, so it also makes sense to do those first, although the Merkle tree check should certainly be done before any transaction checks.

Reply:
I'll add a new commit to move these around.

Reply:
Fixed in eb9ad421